src/index.py

```diff
@@ -79,3 +96,260 @@ def contents_from_page(page: str) -> str:
     """
     contents = page.split(NAVIGATION_TABLE_START)
     return contents[0]
+
+
+class _ParsedListItem(typing.NamedTuple):
+    """Represents a parsed item in the contents table.
+
+    Attrs:
+        whitespace_count: The number of leading whitespace characters
+        reference_title: The name of the reference
+        reference_value: The link to the referenced item
+        rank: The number of preceding elements in the list
+    """
+
+    whitespace_count: int
+    reference_title: str
+    reference_value: str
+    rank: int
+
+
+def _parse_item_from_line(line: str, rank: int) -> _ParsedListItem:
+    """Parse an index list item from a contents line.
+
+    Args:
+        line: The contents line to parse.
+        rank: The number of previous items.
+
+    Returns:
+        The parsed content item.
+
+    Raises:
+        InputError:
+            - When an item is malformed.
+            - When the first item has leading whitespace.
+    """
+    match = _ITEM_PATTERN.match(line)
+
+    if match is None:
+        raise InputError(
+            f"An item in the contents of the index file at {DOCUMENTATION_INDEX_FILENAME} is "
+            f"invalid, {line=!r}, expecting regex: {_ITEM}"
+        )
+
+    whitespace_count = len(match.group(1))
+
+    if whitespace_count != 0 and rank == 0:
+        raise InputError(
+            f"An item in the contents of the index file at {DOCUMENTATION_INDEX_FILENAME} is "
+            f"invalid, {line=!r}, expecting the first line not to have any leading whitespace"
+        )
+
+    reference_title = match.group(7)
+    reference_value = match.group(8)
+
+    return _ParsedListItem(
+        whitespace_count=whitespace_count,
+        reference_title=reference_title,
+        reference_value=reference_value,
+        rank=rank,
+    )
+
+
+def _remove_contents(lines: typing.Iterable[str]) -> typing.Iterator[str]:
+    """Remove the contents section lines from the index contents.
+
+    Args:
+        lines: The lines of the index file.
+
+    Yields:
+        All lines except the lines of the contents section.
+    """
+    contents_encountered = False
+    drop_lines = False
+    for line in lines:
+        if not contents_encountered and line.lower() == CONTENTS_HEADER:
+            contents_encountered = True
+            drop_lines = True
+        elif line.startswith(CONTENTS_END_LINE_PREFIX):
```

Comment:
Is this needed?

Reply:
This generator should yield all the lines from the `index.md` file expect the contents section which can be anywhere in the file. So these two conditionals first disable yielding lines when the contents section starts and then re-enable yielding when it ends. Does that make sense?